### PR TITLE
speedup SnapshotStore.get

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
In contentious case, many many rollbacks may exist for a single key. The commit informations for this key is like that:
```
k vn rollback ts = n
…
k v2 rollback ts = 2
k v1 commit ts = 1
```
When use ts = n+1 to get k, currently we will create iterator and call seek for each version. 
In this case actually we only need create one iterator and use next to iterate all versions.

I test `SnapshotStore.get` on a key with 10000 rollbacks and 1 commit version, the cost time reduced by nearly 80%. I have to remind that in the above test all contents are in RocksDB's Memtable, if the dataset is larger, the positive effect of this pr is much more huge.
```
branch: txn get elapsed Duration { secs: 0, nanos: 44372872 }
master: txn get elapsed Duration { secs: 0, nanos: 209389204 }
```

BTW, 3 LGTM is needed.

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
@wentaoxu PTAL

**Comment by siddontang** [comment]
it is better than we can merge https://github.com/pingcap/tikv/pull/2597 at first, then we can bench this directly.

**Comment by disksing** [comment]
Seems this patch also disables the prefix bloom filter for `seek_write`, may have side effect on `insert`.
https://github.com/pingcap/tikv/blob/f0b80c49066bf1786d2cb26e1a69f8fd027c10bc/src/storage/mvcc/reader.rs#L153-L173

**Comment by zhangjinpeng1987** [comment]
@disksing This pr will not affect prewrite, rollback or commit, please see https://github.com/pingcap/tikv/blob/master/src/storage/txn/scheduler.rs#L706

**Comment by BusyJay** [comment]
Can the iterator be reused even when prefix bloom filter is enabled?

**Comment by zhangjinpeng1987** [comment]
I think the most elegant solution is only keep the latest rollback whose ts is largest.

**Comment by wentaoxu** [comment]
@zhangjinpeng1987 I agree with you, but it is dangerous. Maybe we can only keep rollback record of the primary key? @disksing

**Comment by wentaoxu** [comment]
Maybe worse performance for normal cases with little rollback？ I think it needs more bench.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

File: src/storage/txn/store.rs [hunk 1]
```diff
@@ -90,6 +92,7 @@ impl SnapshotStore {
             lower_bound,
             upper_bound,
             self.isolation_level,
+            false,
```
Comment (AndreMouche) [inline src/storage/txn/store.rs]: Could we use `use_prefix_seek` in `scanner`?

File: src/storage/mvcc/reader.rs [hunk 2]
```diff
@@ -158,7 +162,13 @@ impl MvccReader {
     ) -> Result<Option<(u64, Write)>> {
         if self.scan_mode.is_some() {
             if self.write_cursor.is_none() {
-                let iter_opt = IterOption::new(None, None, self.fill_cache);
+                let iter_opt = if self.use_prefix_seek {
+                    IterOption::default()
+                        .use_prefix_seek()
+                        .set_prefix_same_as_start(true)
```
Comment (disksing) [inline src/storage/mvcc/reader.rs]: So if we enable _use_prefix_seek_, the _fill_cache_ will be ignored?

File: src/storage/txn/store.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/storage/txn/store.rs]: We can't.

File: src/storage/txn/store.rs [hunk 3]
```diff
@@ -279,6 +313,31 @@ mod test {
...
+        store.must_rollback(k, 5);
+        store.must_prewrite_put(k, v4, k, 7);
+        store.must_rollback(k, 7);
+
+        // use start_ts = 21 to get, travel all rollback marks and get committed value v1.
+        store.refresh_snapshot();
+        let snapshot_store = store.store();
+        let mut statistics = Statistics::default();
+        let data = snapshot_store.get(&make_key(k), &mut statistics).unwrap();
+        assert_eq!(data.unwrap(), v1);
```
Comment (disksing) [inline src/storage/txn/store.rs]: Can we use `statistics` to check the _get_ and _next_ op count?

**Comment by disksing** [comment]
LGTM.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL
3 LGTM is required.

**Comment by AndreMouche** [comment]
Any update? Or Would this PR be merged?

**Comment by breeswish** [comment]
LGTM

**Comment by BusyJay** [comment]
Any updates on this?

**Comment by zhangjinpeng1987** [comment]
@BusyJay @breeswish I don't think this solution is good enough, i think compress multiple continuous rollback versions into one or two rollback is a better solution, i'm working on it.

**Comment by zhangjinpeng1987** [comment]
https://github.com/pingcap/tikv/pull/3290 is a better solution.

**Pull request closed without merging**